[MODEL] Nanochat implementation

burtenshaw opened the pull request:
This PR introduces Andrej Karpathy's `nanochat` to transformers.

The transformers compatible weights are currently available in two places:

- A [pr](https://huggingface.co/karpathy/nanochat-d32/discussions/1) on Karpathy's repo (large model)
- A [repo](https://huggingface.co/nanochat-students/nanochat-d20) on the community org  (small model)

# ToDo

- CI is failing dues to tests for `SwiftFormerForQuestionAnswering` etc.

burtenshaw committed:
first draft on modelling

burtenshaw committed:
add modelling to auto

burtenshaw committed:
add to init

burtenshaw committed:
[WIP] add scripts for conversion

burtenshaw committed:
hack at attention and rotary using logit comparison unitl it works

burtenshaw committed:
update conversion script

burtenshaw committed:
fix test

burtenshaw committed:
tody up decoding inputs

burtenshaw committed:
fix all naming nanogpt >> nanochat

burtenshaw committed:
add return as tensors

burtenshaw committed:
use batch encoding

xenova commented on src/transformers/models/nanochat/tokenizer_nanochat.py:
I know the PR is still in draft, but we might not need this file. Most of the chat template logic should be able to be moved to the chat_template.jinja file.

Also, if we do need this file, it should be named `tokenization_nanochat.py`

burtenshaw committed:
include tokenizer conversion in the conversion script

burtenshaw committed:
add repo revision to vocab pull

burtenshaw committed:
move attention into func and add kward to all signatures (for vllm)

burtenshaw committed:
rename everything NanoGPT >> NanoChat

burtenshaw committed:
improve docstrings in modelling code

burtenshaw committed:
run ruff

burtenshaw committed:
delete working test file

burtenshaw committed:
remove custom tokenizer

burtenshaw committed:
protect imports with lazy

burtenshaw committed:
removed unused variables

burtenshaw committed:
format config

burtenshaw committed:
sort auto model configs

burtenshaw committed:
make sure nanochat config is in all (fixes vllm)

burtenshaw committed:
integrate tests

burtenshaw committed:
remove custom tokenizer from autos

burtenshaw committed:
formate tests

burtenshaw committed:
add documentation for nanochat

burtenshaw committed:
update toc with nanochat

burtenshaw committed:
Merge branch 'main' into nanochat-implementation

xenova commented on src/transformers/models/nanochat/configuration_nanochat.py:
```diff
@@ -0,0 +1,100 @@
+from transformers.configuration_utils import PretrainedConfig
+from transformers.utils import logging
```
```suggestion
from ...configuration_utils import PretrainedConfig
from ...utils import logging
```

burtenshaw committed:
lose some attention notes

vasqu commented on scripts/convert_nanogpt_checkpoint.py:
Conversion scripts usually also go under the model, so `models/nanochat/convert_....`

vasqu commented on src/transformers/models/nanochat/configuration_nanochat.py:
Dont forget the licences

vasqu commented on src/transformers/models/nanochat/configuration_nanochat.py:
```diff
@@ -0,0 +1,100 @@
+
+
+class NanoChatConfig(PretrainedConfig):
+    r"""Configuration for the NanoChat model."""
```
Probably aware but docs please

vasqu commented on src/transformers/models/nanochat/configuration_nanochat.py:
```diff
@@ -0,0 +1,100 @@
+            eos_token_id=eos_token_id,
+            pad_token_id=pad_token_id,
+            **kwargs,
+        )
```
Should be at the end

vasqu commented on src/transformers/models/nanochat/modeling_nanochat.py:
```diff
@@ -0,0 +1,497 @@
+logger = logging.get_logger(__name__)
+
+
+def _apply_rotary_pos_emb(
```
I will add comments step by step explaining why this is essentially Llama RoPE

vasqu commented on src/transformers/models/nanochat/modeling_nanochat.py:
```diff
@@ -0,0 +1,497 @@
+    k1, k2 = key[..., :d], key[..., d:]
+
+    query_rot = torch.cat([q1 * cos + q2 * sin, q1 * (-sin) + q2 * cos], dim=-1)
+    key_rot = torch.cat([k1 * cos + k2 * sin, k1 * (-sin) + k2 * cos], dim=-1)
```
A bit more complicated and we need a slight adjustment in our `rotate_half`. In Llama we rotate with https://github.com/huggingface/transformers/blob/fe11cbb808b4301399240e40c5cb6cca9bb00d4d/src/transformers/models/llama/modeling_llama.py#L113
and add with https://github.com/huggingface/transformers/blob/fe11cbb808b4301399240e40c5cb6cca9bb00d4d/src/transformers/models/llama/modeling_llama.py#L138
Mind that our prepared sin/cos are `head_dim`, not `head_dim//2`, by duplicating https://github.com/huggingface/transformers/blob/fe11cbb808b4301399240e40c5cb6cca9bb00d4d/src/transformers/models/llama/modeling_llama.py#L102

Mathematically, with Llama we get:
```
[q1 * cos - q2 * sin, q1 * sin + q2 * cos]
```
Not that we rotate correctly but the minus/plus signs don't add up. For this we need to change the rotation slightly from https://github.com/huggingface/transformers/blob/fe11cbb808b4301399240e40c5cb6cca9bb00d4d/src/transformers/models/llama/modeling_llama.py#L113
to `torch.cat((x2, -x1), dim=-1)` so that we go to 
```
[q1 * cos + q2 * sin, -(q1 * sin) + q2 * cos]
```
which is the same as here

vasqu commented on src/transformers/models/nanochat/modeling_nanochat.py:
```diff
@@ -0,0 +1,497 @@
+        positions = torch.arange(self.config.max_position_embeddings, device=device, dtype=torch.float32)
+        # Stride the channels (only even channels, head_dim//2 frequencies)
+        freqs = torch.arange(0, head_dim, 2, dtype=torch.float32, device=device)
+        inv_freq = 1.0 / (self.config.rope_theta ** (freqs / head_dim))
```
The same as default rope (in llama) https://github.com/huggingface/transformers/blob/bdbc2d037b1751c71b30a488f0a168deabef9a90/src/transformers/modeling_rope_utils.py#L125-L131

vasqu commented on src/transformers/models/nanochat/modeling_nanochat.py:
```diff
@@ -0,0 +1,497 @@
+        angles = torch.einsum("i,j->ij", positions, inv_freq)
+        # Compute cos and sin, add batch and head dims for broadcasting
+        cos = angles.cos()[None, :, None, :]  # [1, seq_len, 1, head_dim//2]
+        sin = angles.sin()[None, :, None, :]  # [1, seq_len, 1, head_dim//2]
```
We have `[bsz, seq_len, head_dim]` instead by duplicating https://github.com/huggingface/transformers/blob/fe11cbb808b4301399240e40c5cb6cca9bb00d4d/src/transformers/models/llama/modeling_llama.py#L102
and the aforementioned extension of the outer product to 2D (positions)

vasqu commented on src/transformers/models/nanochat/modeling_nanochat.py:
```diff
@@ -0,0 +1,497 @@
+    return hidden_states.reshape(batch, num_key_value_heads * n_rep, seq_len, head_dim)
+
+
+def eager_attention_forward(
```
https://github.com/huggingface/transformers/blob/fe11cbb808b4301399240e40c5cb6cca9bb00d4d/src/transformers/models/llama/modeling_llama.py#L171

vasqu commented on src/transformers/models/nanochat/modeling_nanochat.py:
```diff
@@ -0,0 +1,497 @@
+
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        hidden_states = self.fc(hidden_states)
+        hidden_states = F.relu(hidden_states).square()
```
We have a relu² act here https://github.com/huggingface/transformers/blob/bdbc2d037b1751c71b30a488f0a168deabef9a90/src/transformers/activations.py#L206

We should be able to infer this with something like https://github.com/huggingface/transformers/blob/fe11cbb808b4301399240e40c5cb6cca9bb00d4d/src/transformers/models/llama/modeling_llama.py#L152

And there is probably a model which does this then from which we can possibly inherit

vasqu commented on src/transformers/models/nanochat/modeling_nanochat.py:
```diff
@@ -0,0 +1,497 @@
+        **kwargs: Unpack[TransformersKwargs],
+    ) -> tuple[torch.Tensor, Optional[torch.Tensor]]:
+        residual = hidden_states
+        hidden_states = F.rms_norm(hidden_states, (hidden_states.size(-1),), eps=self.config.rms_norm_eps)
```
See my comments in the attention module, we should avoid using functional here, and make this a module instead, should also allow for easier inheritance

vasqu commented on src/transformers/models/nanochat/modeling_nanochat.py:
```diff
@@ -0,0 +1,497 @@
+    ) -> tuple[torch.Tensor, Optional[torch.Tensor]]:
+        residual = hidden_states
+        hidden_states = F.rms_norm(hidden_states, (hidden_states.size(-1),), eps=self.config.rms_norm_eps)
+        hidden_states, self_attn_weights = self.self_attn(
```
```suggestion
        hidden_states, _ = self.self_attn(
```
Llama style with https://github.com/huggingface/transformers/blob/fe11cbb808b4301399240e40c5cb6cca9bb00d4d/src/transformers/models/llama/modeling_llama.py#L325-L328

this will handle everything for you, kwargs, collecting the tensors etc

vasqu commented on src/transformers/models/nanochat/modeling_nanochat.py:
```diff
@@ -0,0 +1,497 @@
+    config_class = NanoChatConfig
+    base_model_prefix = "model"
+    _no_split_modules = ["NanoChatDecoderLayer"]
+    _supports_attention_backend = True
```
We should have all flags enabled here https://github.com/huggingface/transformers/blob/fe11cbb808b4301399240e40c5cb6cca9bb00d4d/src/transformers/models/llama/modeling_llama.py#L316-L324

Currently, it seems only eager is run then + the record attribute as a reminder (handles output_xxx)

https://github.com/huggingface/transformers/blob/fe11cbb808b4301399240e40c5cb6cca9bb00d4d/src/transformers/models/llama/modeling_llama.py#L325-L328

burtenshaw replied:
> Currently, it seems only eager attention is run then + the record attribute as a reminder (handles output_xxx)

Sorry. I don't fully understand this. Can you check [16a0207d9e1fedf0a9f0445a60bddcfbe722d02b](https://github.com/huggingface/transformers/pull/41634/commits/16a0207d9e1fedf0a9f0445a60bddcfbe722d02b) please.

vasqu replied:
I think you still need to convert to the modeling file, i.e. something along `python utils/modular_model_converter.py nanochat`. But yes, that looks correct. 

- Each flag indicates that another attn flavor can be run --> no flags == only eager attention can be run, no other more efficient variations such as SDPA or Flash Attn.
- The record flag allows avoiding collecting all these tensors manually, i.e. no more tuples/lists to collect manually + the output_xxx kwargs also handled automatically then.

vasqu commented on src/transformers/models/nanochat/configuration_nanochat.py:
```diff
@@ -0,0 +1,100 @@
+        "num_hidden_layers": "n_layer",
+        "num_attention_heads": "n_head",
+    }
+
```
Just a random point but it might be nice to have a tp/pp plan if the model is big https://github.com/huggingface/transformers/blob/bdbc2d037b1751c71b30a488f0a168deabef9a90/src/transformers/models/llama/configuration_llama.py#L146-L159

vasqu reviewed (commented):
An initial review, I went a bit into the deep end with the RoPE but I hope I could explain what's going on :sweat_smile: 

The biggest tasks are 
- modular
- RoPE
- `_can_record` to properly support some features

burtenshaw committed:
remove excess configuration

burtenshaw committed:
move init

burtenshaw committed:
move script into modeling dir

burtenshaw committed:
switch to modular and inheret llama

burtenshaw committed:
add causal flag

burtenshaw committed:
remove excess  and just set params in model config

burtenshaw committed:
liscence

burtenshaw committed:
update model flags

burtenshaw committed:
re use init weights

burtenshaw committed:
lose functionals from the MLP

burtenshaw committed:
use rmsnorm module instead of functional

burtenshaw committed:
inherit rms from NanoChatRMSNorm

burtenshaw committed:
run modular convert and update modeling

burtenshaw committed:
update docs

burtenshaw committed:
remove excess changes in auto

burtenshaw committed:
Update src/transformers/models/nanochat/configuration_nanochat.py

Co-authored-by: Joshua Lochner <admin@xenova.com>

burtenshaw committed:
add a basic tp/pp plan

burtenshaw committed:
Merge branch 'nanochat-implementation' of https://github.com/huggingface/transformers into nanochat-implementation

burtenshaw committed:
use cleanup util for tests

burtenshaw committed:
update conversion script

burtenshaw committed:
add docstring to nanochat config

burtenshaw committed:
update auto config with NanoChatForCausalLM

burtenshaw committed:
remove forward snippet

burtenshaw committed:
fix auto modeling

burtenshaw committed:
delete excess modelling file

burtenshaw committed:
delete excess modelling file and regenerate modeling

burtenshaw committed:
Merge branch 'nanochat-implementation' of https://github.com/huggingface/transformers into nanochat-implementation

burtenshaw committed:
revert changes and use less llama modules

vasqu commented on src/transformers/models/nanochat/modeling_nanochat.py:
```diff
@@ -343,18 +371,13 @@ def _init_weights(self, module: nn.Module) -> None:
         self.vocab_size = config.vocab_size
 
         self.embed_tokens = nn.Embedding(config.vocab_size, config.hidden_size, self.padding_idx)
+        self.initial_norm = NanoChatRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
```
Could it be that this was the regression, i.e. a norm was missing that was there before?

vasqu commented on src/transformers/models/nanochat/modeling_nanochat.py:
```diff
@@ -406,45 +438,63 @@ def forward(
 
         hidden_states = inputs_embeds
         position_embeddings = self.rotary_emb(hidden_states, position_ids)
+        hidden_states = self.initial_norm(hidden_states)
```
This seems like the only functional real change

vasqu commented on src/transformers/models/nanochat/modeling_nanochat.py:
```diff
@@ -65,38 +90,31 @@ def __init__(self, config: NanoChatConfig, device=None):
-    Overrides __init__ to match NanoChat's API with hidden_size parameter.
-    """
+        # Add extra dimension for NanoChat's broadcasting: [batch, seq_len] -> [batch, seq_len, 1, head_dim//2]
+        return cos.unsqueeze(2).to(dtype=x.dtype), sin.unsqueeze(2).to(dtype=x.dtype)
```
We later squeeze and unsqueeze again. Imo, the changes to the RoPE seems unnecessary to me :thinking:

vasqu reviewed (commented):
Just some quick comments because I see your comment about a regression. We should really not have to change the RoPE by that much. It rather seems to me that a rms normalization that was missing?

I will take a proper look on monday

burtenshaw committed:
use llama rope not custom

vasqu commented on src/transformers/models/nanochat/modular_nanochat.py:
```diff
@@ -0,0 +1,431 @@
+
+    def __init__(self, hidden_size, eps=1e-6):
+        super().__init__(eps=eps)
+        self.hidden_size = hidden_size
```
Can be removed, no?

vasqu replied:
Let's change this in the inits of the other modules instead. It's weird to have hidden_size when it's unused

vasqu commented on src/transformers/models/nanochat/modular_nanochat.py:
```diff
@@ -0,0 +1,431 @@
+        # Project the input to get queries, keys, and values [batch, num_heads, seq_len, head_dim]
+        query_states = (
+            self.q_proj(hidden_states).view(batch, seq_len, self.num_heads, self.head_dim).transpose(1, 2).contiguous()
+        )
```
Also the contiguous is unnecessary, we call it if it's necessary in our attention versions

vasqu commented on src/transformers/models/nanochat/modular_nanochat.py:
```diff
@@ -0,0 +1,431 @@
+
+        # Apply KV cache: insert current k,v into cache, get the full view so far
+        if past_key_values is not None:
+            cache_kwargs = {"cache_position": cache_position}
```
https://github.com/huggingface/transformers/blob/ce4ffeeb6c39a9e7e2e62e23606d0b83b9a49b96/src/transformers/models/qwen3/modeling_qwen3.py#L274

vasqu commented on src/transformers/models/nanochat/modular_nanochat.py:
```diff
@@ -0,0 +1,431 @@
+        return attn_output, attn_weights
+
+
+class NanoChatMLP(nn.Module):
```
Inheritable from https://github.com/huggingface/transformers/blob/818f7f10e40cbceb3ce8a8b9a1f174f6b9181da2/src/transformers/models/clip/modeling_clip.py#L354

Just needs renames of the projections

vasqu commented on src/transformers/models/nanochat/modular_nanochat.py:
```diff
@@ -0,0 +1,431 @@
+        **kwargs: Unpack[TransformersKwargs],
+    ) -> BaseModelOutputWithPast:
+        output_attentions = kwargs.get("output_attentions", False)
+        output_hidden_states = kwargs.get("output_hidden_states", False)
```
We don't need this, please copy https://github.com/huggingface/transformers/blob/818f7f10e40cbceb3ce8a8b9a1f174f6b9181da2/src/transformers/models/llama/modeling_llama.py#L378-L436

and only add the initial norm before the layers start

vasqu reviewed (commented):
We are not using modular to its fullest, I added some comments where we could apply modular:
- RoPE change rotate half again to align with our attention modules' implementation
- Attention -> Qwen3 Attention; forward might need one adjustment = order of ops (rope, rms norm) in the forward
- MLP -> Clip MLP
- Decoder Layer -> Llama Decoder Layer
- PreTrainedModel -> Llama PreTrainedModel (with different init weights)
- Model -> Llama (at least for the init)
- CausalLM -> Llama (mostly the same except the softcaping in the forward)

I tried to find the differences, but maybe one or two changes are additionally needed, e.g. some attribute has a different name, a norm needs a diffferent init etc

vasqu commented on src/transformers/models/nanochat/modular_nanochat.py:
```diff
@@ -0,0 +1,431 @@
+
+
+@auto_docstring
+class NanoChatModel(NanoChatPreTrainedModel):
```
(only init, forward needs to change see below)

burtenshaw committed:
cut down on notes and docstrings

burtenshaw committed:
revert again to custom rope but this time using Qwen3 attention

burtenshaw committed:
tidy notes in attention module

burtenshaw committed:
update cache kwargs with sin and cos

burtenshaw committed:
inheret from CLIPMLP and overwrite projections

burtenshaw committed:
inheret llama decode layer

burtenshaw committed:
re use more from LlamaPreTrainedModel

burtenshaw committed:
inherit LlamaModel and copy forward except for initial_norm

burtenshaw committed:
inheret more from LlamaForCausalLM

vasqu committed:
more modular friendly

vasqu committed:
Merge remote-tracking branch 'upstream/main' into nanochat-implementation

vasqu committed:
adjust to changes in main

vasqu reviewed (commented):
Thanks for bearing with me @burtenshaw, I know I'm petty on a lot things here :sweat_smile: I added a few changes to make a bit more use out of modular. The nit was confusing because main diverged with the current main of this branch, you can now see what I do in the modular class.

I added a few TODOs and FIXMEs which will likely need to be addressed. These are mostly related to check for the conversion script and the config. Lmk if something is unclear

vasqu committed:
last nits

vasqu committed:
apparently we do this now lol

vasqu committed:
mustve been a blank smthn smthn

burtenshaw committed:
update conversion script to use original config

burtenshaw committed:
fix attetion_bias everywhere

burtenshaw committed:
resolve todos in config

burtenshaw committed:
todos: layer names and repo id in modular

burtenshaw committed:
last todos in config and modelling

burtenshaw committed:
usng karapthy repo

vasqu committed:
inherit from gemma instead

vasqu commented on src/transformers/models/nanochat/convert_nanochat_checkpoints.py:
```diff
@@ -0,0 +1,311 @@
+                elif key == "attention_bias" and "qkv_bias" in extra_config:
+                    config_kwargs[key] = extra_config["qkv_bias"]
+                elif key == "final_logit_softcapping" and "logits_soft_cap":
+                    config_kwargs[key] = extra_config["logits_soft_cap"]
```
This is to align with gemma2 so we can use modular.

vasqu commented on tests/models/nanochat/test_modeling_nanochat.py:
```diff
@@ -0,0 +1,207 @@
+
+        # The model should generate a reasonable response (with greedy decoding this is deterministic)
+        # Expected: "The capital of France is Paris."
+        self.assertIn("Paris", generated_text)
```
Let's generate a number of tokens and compare against the sequence instead. 1 Token is a bit wonky. 

You can look at Llama for example, imo something along 32 new tokens would suffice. Same goes for the other generation tests

vasqu reviewed (commented):
Finish line! I noticed that the soft capping is also done in gemma models, so I changed a few things there; if it doesnt work out I'll revert - no need to do anything on your side.

Remaining things to check:
- I didnt break the conversion script
- Expand the generation tests to have a longer sequence to check against (e.g. 32 tokens)
- Fix integration tests (?), see https://github.com/huggingface/transformers/actions/runs/18712904195/job/53365510466

For the integrations test, we use a10 gpus. If you confirm that everything works on your side, I can just update the values if needed

vasqu committed:
fix vaultgemma

vasqu committed:
Merge branch 'main' into nanochat-implementation

vasqu committed:
ci common

vasqu commented:
run-slow: gemma3

burtenshaw committed:
do rope config after super init

burtenshaw committed:
update logit values in tests

burtenshaw committed:
update logit tests one last time

burtenshaw committed:
update and format conversion script

burtenshaw committed:
Merge branch 'nanochat-implementation' of https://github.com/huggingface/transformers into nanochat-implementation

vasqu commented on tests/models/nanochat/test_modeling_nanochat.py:
```diff
@@ -0,0 +1,207 @@
+        EXPECTED_SLICE = torch.tensor(
+            [-12.4375, -13.1875, -12.875, -13.1875, -13.1875, -13.1875, -13.1875, -13.1875, -12.0, -1.6172]
+        )
+        torch.testing.assert_close(logits[0, 0, :10], EXPECTED_SLICE, rtol=1e-3, atol=0.1)
```
Let's keep the atols as low as possible; 1e-1 is too high. I can update on our runners if everything on your side is good

burtenshaw committed:
update all logits again

burtenshaw committed:
update test tollerence

vasqu committed:
fix tests

vasqu committed:
auto tokenizer

vasqu commented on src/transformers/models/auto/tokenization_auto.py:
```diff
@@ -468,6 +468,7 @@
         ("musicgen_melody", ("T5Tokenizer", "T5TokenizerFast" if is_tokenizers_available() else None)),
         ("mvp", ("MvpTokenizer", "MvpTokenizerFast" if is_tokenizers_available() else None)),
         ("myt5", ("MyT5Tokenizer", None)),
+        ("nanochat", (None, "PreTrainedTokenizerFast" if is_tokenizers_available() else None)),
```
Added to auto, just to be sure can you cross check if that's correct

vasqu commented on tests/models/nanochat/test_modeling_nanochat.py:
```diff
@@ -0,0 +1,233 @@
+        ]
+
+        EXPECTED_TEXT_COMPLETION = [
+            "The capital of France is Paris.PatrickPatrickPatrickPatrickPatrickPatrickPatrickPatrickPatrickPatrickPatrickPatrick",
```
It's due to the generation config having wrong token ids https://huggingface.co/nanochat-students/nanochat-d20/blob/main/generation_config.json

TODO: update the config

vasqu committed:
fix test, to be updated upstream on the hub

burtenshaw committed:
Merge branch 'main' into nanochat-implementation

vasqu commented on src/transformers/models/gemma2/modular_gemma2.py:
For visibility, these changes here and in the other gemma models are intentional:
- It wasn't properly using our features or introduced manual stuff we did before which is no longer necessary
- Enables nanochat to inherit from gemma2 properly

vasqu committed:
Merge branch 'main' into nanochat-implementation

ydshieh commented:
run-slow: nanochat

vasqu commented:
cc @ArthurZucker @Cyrilvallez for core maintainer when you have time

ydshieh commented:
Have to resolve the conflicts :-)

vasqu committed:
Merge branch 'main' into nanochat-implementation

vasqu committed:
fix

burtenshaw committed:
update conversion script to deal with new d34 model

vasqu committed:
Merge branch 'main' into nanochat-implementation

Cyrilvallez committed:
Merge branch 'main' into nanochat-implementation

Cyrilvallez commented on src/transformers/models/nanochat/modular_nanochat.py:
```diff
@@ -0,0 +1,247 @@
+                    param,
+                    mean=0.0,
+                    std=self.config.initializer_range / math.sqrt(2 * self.config.num_hidden_layers),
+                )
```
We absolutely cannot do that unfortunately, as we will iterate over all params for every module. If we need to change the default scheme, we need to rewrite _init_weights completely. Here, you could add a check isinstance(module, NanoChatAttention)`, and init correctly the o_proj from this

Cyrilvallez reviewed (commented):
Amazing work! Super clean, and you also refactored gemmas super nicely 🤗🤗🚀
It's really almost ready to be merged, only 2 frictions points are the `_init_weights`, and the tp_plan currently! They need to be changed a bit for correctness/efficiency! Let me know if something is unclear about that!

burtenshaw committed:
fix eval in docs

burtenshaw committed:
Apply suggestions from code review

Co-authored-by: Cyril Vallez <cyril.vallez@huggingface.co>

burtenshaw committed:
no separate initial_norm and do not iter over module

burtenshaw committed:
update _tp_plan

burtenshaw committed:
update modeling with changes

burtenshaw commented:
> Amazing work! Super clean, and you also refactored gemmas super nicely 🤗🤗🚀 It's really almost ready to be merged, only 2 frictions points are the `_init_weights`, and the tp_plan currently! They need to be changed a bit for correctness/efficiency! Let me know if something is unclear about that!

@Cyrilvallez Thanks for the review! I've responded to all those changes and added suggestions.

vasqu commented:
run-slow: nanochat

burtenshaw committed:
Merge branch 'main' into nanochat-implementation

Cyrilvallez committed:
style

Cyrilvallez committed:
fix tp_plan

Cyrilvallez reviewed (approved):
Alright, amazing! Thanks a lot! I pushed the last details and merging it now!! Congrats! 🤗🤗

Cyrilvallez committed:
oupppss this one almost slipped through nice I checked
